File: pages/refactorings/extract-method/01-order-total/Before.scala
// Total of an order: the lines summed, less a percentage discount, plus tax.
object Before:
  case class Line(unitPrice: Int, quantity: Int)
  case class Order(items: List[Line], discountPct: Int, taxPct: Int)

  def total(o: Order): Int =
    val subtotal   = o.items.map(l => l.unitPrice * l.quantity).sum
    val discounted = subtotal - subtotal * o.discountPct / 100
    discounted + discounted * o.taxPct / 100

File: pages/refactorings/extract-method/01-order-total/Before.hs
-- Total of an order: the lines summed, less a percentage discount, plus tax.
module Before where

data Line  = Line  { unitPrice :: Int, quantity :: Int }
data Order = Order { items :: [Line], discountPct :: Int, taxPct :: Int }

total :: Order -> Int
total o = discounted + discounted * taxPct o `div` 100
  where
    subtotal   = sum [unitPrice l * quantity l | l <- items o]
    discounted = subtotal - subtotal * discountPct o `div` 100
